wasi outbount http: add request config

flavio opened the pull request:
This is an attempt to port https://github.com/deislabs/wasi-experimental-http/pull/83 to use witx as a way to define the configuration objects.

flavio committed:
Add HTTP request config to witx definitions

Allow outbound HTTP request to provide a configuration.

The purpose of the configuration is to allow fine tuning of values like:

  * add extra CA to the list of the trusted ones
  * support mutual TLS authentication

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio committed:
Update witx rev

This is required to fix a compilation error related with wasi cache

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio committed:
WIP: make use of the http request config

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio commented:
I'm currently having some issues, I fear there's something broken with the wit-bindgen-wamstime...

This is the error I get when I perform a `cargo build` inside of the `crates/http-wasmtime` directory:

```shell
error[E0277]: the trait bound `[closure@crates/http-wasmtime/src/lib.rs:10:1: 10:70]: IntoFunc<T, _, _>` is not satisfied
   --> crates/http-wasmtime/src/lib.rs:10:1
    |
10  | wit_bindgen_wasmtime::export!("wit/ephemeral/wasi-outbound-http.wit");
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `IntoFunc<T, _, _>` is not implemented for `[closure@crates/http-wasmtime/src/lib.rs:10:1: 10:70]`
    |
note: required by a bound in `Linker::<T>::func_wrap`
   --> /home/flavio/.cargo/registry/src/github.com-1ecc6299db9ec823/wasmtime-0.33.0/src/linker.rs:432:20
    |
432 |         func: impl IntoFunc<T, Params, Args>,
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Linker::<T>::func_wrap`
    = note: this error originates in the macro `wit_bindgen_wasmtime::export` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `wasi-outbound-http-wasmtime` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
╭─ flavio@ganymede  ~/hacking/wasm/wasi-experimental-toolkit/crates/http-wasmtime   main ● ⍟3                                                                                                                          101 ↵  09:47:14 
╰─ cargo build
   Compiling wasi-outbound-http-wasmtime v0.1.0 (/home/flavio/hacking/wasm/wasi-experimental-toolkit/crates/http-wasmtime)
error[E0277]: the trait bound `[closure@crates/http-wasmtime/src/lib.rs:10:1: 10:70]: IntoFunc<T, _, _>` is not satisfied
   --> crates/http-wasmtime/src/lib.rs:10:1
    |
10  | wit_bindgen_wasmtime::export!("wit/ephemeral/wasi-outbound-http.wit");
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `IntoFunc<T, _, _>` is not implemented for `[closure@crates/http-wasmtime/src/lib.rs:10:1: 10:70]`
    |
note: required by a bound in `Linker::<T>::func_wrap`
   --> /home/flavio/.cargo/registry/src/github.com-1ecc6299db9ec823/wasmtime-0.33.0/src/linker.rs:432:20
    |
432 |         func: impl IntoFunc<T, Params, Args>,
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Linker::<T>::func_wrap`
    = note: this error originates in the macro `wit_bindgen_wasmtime::export` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `wasi-outbound-http-wasmtime` due to previous error
```

It turns out that removing the `identity` object from the `request-config` record fixes the build. I tried to figure out what is wrong with the `identity` record, but I couldn't understand that.

I didn't look into the source codes of `wit-bindgen-wamstime` yet...

@radu-matei: have you ever ran into something similar?

radu-matei commented on Cargo.toml:
```diff
@@ -22,4 +22,4 @@
     wasi-common                 = "0.33"
     wasi-cap-std-sync           = "0.33"
-    wit-bindgen-wasmtime        = { git = "https://github.com/bytecodealliance/wit-bindgen", rev = "2e654dc82b7f9331719ba617a36ed5967b2aecb0" }
+    wit-bindgen-wasmtime        = { git = "https://github.com/bytecodealliance/wit-bindgen", rev = "8fdf702982ab141c63d5b1885994306c77dbc41d" }
```
First, you are only bumping the revision of `wit-bindgen` here, whereas this crate is imported in the top level `Cargo.toml` — so the first thing I'd do is either bump all, or none.

radu-matei commented:
I haven't ran in this before, but managed to reproduce it with a minimal example in https://github.com/bytecodealliance/wit-bindgen/issues/134.

radu-matei commented:
This most likely turns out to be because of the high number of parameters in the generated canonical ABI (which lowers the types in WIT to Wasm data types).

I think that for the time being, we might need to look at limiting the records / fields in the records to the stay withing the 16 parameters limit.

flavio committed:
More WIP: workaround for witx limitation

The previous code didn't build becuse of a witx limitation,
This commit solves the issue by using a different approach.

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio commented:
@radu-matei I finally had some time to look at this issue again. I think I found a workaround for the witx limitation you found.

Can you take a look at the code?

The unit tests are passing, but I would like to test it in a real-world scenario. About that, how is this crate supposed to be consumed?

I'm trying to build a wasmtime host that understands the outbound http functions, however it looks like building is broken:

This is my `Cargo.toml`:

```toml
[package]
name = "krew-wasm"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
wasmtime = "0.33"
wasi-outbound-http-wasmtime = { git = "https://github.com/flavio/wasi-experimental-toolkit", branch = "wasi-outbount-http-add-request-config"}
```

The `main.rs` is just a simple "hello world". I'm not even including this crate and I get the following error:

```
╰─ cargo check
    Checking wasi-outbound-http-wasmtime v0.1.0 (https://github.com/flavio/wasi-experimental-toolkit?branch=wasi-outbount-http-add-request-config#bd44c116)
error: failed to read: wit/ephemeral/wasi-outbound-http.wit
  --> /home/flavio/.cargo/git/checkouts/wasi-experimental-toolkit-39bb3f14a2930fc1/bd44c11/crates/http-wasmtime/src/lib.rs:19:1
   |
19 | wit_bindgen_wasmtime::export!("wit/ephemeral/wasi-outbound-http.wit");
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `wit_bindgen_wasmtime::export` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0432]: unresolved import `wasi_outbound_http`
  --> /home/flavio/.cargo/git/checkouts/wasi-experimental-toolkit-39bb3f14a2930fc1/bd44c11/crates/http-wasmtime/src/lib.rs:12:5
   |
12 | use wasi_outbound_http::*;
   |     ^^^^^^^^^^^^^^^^^^ use of undeclared crate or module `wasi_outbound_http`

# and more
```

it looks like the `wit` file must be part of the `crates/http-wasmtime` directory :thinking: 

I'll make a try during the weekend, now I'm a bit :sleepy:

flavio committed:
Update to latest version of wit-bindgen

Update to the latest version of wit-bindgen, this is required to get
some bug fixes.

As a result of the update the following changes have been done:

  * Update to a more recent version of wasmtime
  * Fix the wasi-ce wit definition: use the new way to escape
  identifiers

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio committed:
wasi-outbound-http: reorganize code

Move the WIT definitions to a dedicated crate. This crate can then be
imported by importers (Wasm modules) and exporters (Wasm hosts).

Prior to this change, all the WIT files were defined inside the root of
the `wasi-experimental-toolkit` crate. That worked fine for the
sub-crates and for the tests because they had direct access to the
files. However that didn't work when trying to consume these files
inside of standalone projects. That's because the WIT files were not
part of the modules checked out by cargo.

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio committed:
wasi-log: reorganize code

Move the WIT definitions to a dedicated crate. This crate can then be
imported by importers (Wasm modules) and exporters (Wasm hosts).

Prior to this change, all the WIT files were defined inside the root of
the `wasi-experimental-toolkit` crate. That worked fine for the
sub-crates and for the tests because they had direct access to the
files. However that didn't work when trying to consume these files
inside of standalone projects. That's because the WIT files were not
part of the modules checked out by cargo.

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio committed:
wasi-cache: reorganize code

Move the WIT definitions to a dedicated crate. This crate can then be
imported by importers (Wasm modules) and exporters (Wasm hosts).

Prior to this change, all the WIT files were defined inside the root of
the `wasi-experimental-toolkit` crate. That worked fine for the
sub-crates and for the tests because they had direct access to the
files. However that didn't work when trying to consume these files
inside of standalone projects. That's because the WIT files were not
part of the modules checked out by cargo.

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio committed:
wasi-ce-http: reorganize code

Move the WIT definitions to a dedicated crate. This crate can then be
imported by importers (Wasm modules) and exporters (Wasm hosts).

Prior to this change, all the WIT files were defined inside the root of
the `wasi-experimental-toolkit` crate. That worked fine for the
sub-crates and for the tests because they had direct access to the
files. However that didn't work when trying to consume these files
inside of standalone projects. That's because the WIT files were not
part of the modules checked out by cargo.

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio committed:
wasi-nn: reorganize code

Move the WIT definitions to a dedicated crate. This crate can then be
imported by importers (Wasm modules) and exporters (Wasm hosts).

Prior to this change, all the WIT files were defined inside the root of
the `wasi-experimental-toolkit` crate. That worked fine for the
sub-crates and for the tests because they had direct access to the
files. However that didn't work when trying to consume these files
inside of standalone projects. That's because the WIT files were not
part of the modules checked out by cargo.

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio committed:
Update list of crates

Add all the new crates that provide WIT definitions

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio committed:
Remove WIT files from root of repo

These files have been moved into the relative `wasi-<something>-def`
crates.

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio commented:
So, after some digging and experimentation I think I found a solution.

The problem was:

  * All the WIT files were part of the root of the repository
  * When writing a Wasm runtime host leveraging one of these functionalities (e.g.: http outbound), the code would only use the crate providing that functionality (`wasi-outbound-http-wasmtime`)

Building the "host runtime" project would then fail because the crate providing the functionality would not build. The build error is caused by the `wit_bindgen_wasmtime::export!("../../wit/ephemeral/wasi-outbound-http.wit")` macro. The macro tries to find the WIT file definition using a relative path, this path cannot be resolved because the WIT files are outside of the crate.

The same problem happens when trying to write a WebAssembly guest that consumes the same WIT file. This time the `wit_bindgen_wasmtime::import!` would fail with the same error.

I think I found a reasonable solution to this problem, this is the approach I'm taking:

  * Create a new library crate that provides the WIT definitions. This crate contains the WIT files. Unfortunately there isn't currently a way for cargo to install assets that are not rust files. To workaround that, the library offers a public static HashMap that has the names of the WIT files as key, and their contents as value
  * All the crates requiring these WIT definitions follow this approach:
    * Introduce a **build-time** dependency against the `wasi-<feature>-defs` crate described above
    * Introduce a `wit/ephemeral` directory, which has a `.gitignore` file that ignores all the `*.wit` files.
    * Introduce a `build.rs` file that dumps the contents of the HashMap described above into `wit/ephemeral/`
    * Rewrite all the `wit_bindgen_wasmtime::export!` and `wit_bindgen_wasmtime::import!` statements to use `wit/ephemeral/file.wit` instead of `../../wit/ephemeral/file.wit`

I had to do this change to all the crates defined inside of the repository. This isn't something I was looking forward, it has been quite some work and I didn't know how you would have reacted to this big change. Hopefully you will be fine with that :crossed_fingers: :smile: 
Looking at how you're using some of these crates inside of `spin`, I think these changes could be useful to you too.

I also had to update to a more recent version of wit-bindgen in order to get the a fix. That triggered an update to a more recent version of wasmtime.

With this change:
  * All the unit test are passing
  * I can build a host runtime without any issue
  * I can build a guest WebAssembly module that leverages one of these libraries.

flavio commented:
I think this is ready for review, I'll fix the history of the repo (I'm referring to the commits with `WIP` subject) before the final merge

radu-matei commented:
Answering specifically, this is used in Spin:

https://github.com/fermyon/spin/blob/e9741e79abb0d6303d7bbbd7dc479b0265b02dc7/crates/engine/src/lib.rs#L106-L108

We hope to be able to pull the crate from here rather than duplicating, but this was the far easier approach for now.

radu-matei commented:
This is awesome, @flavio, thanks a lot for your work!
I will have a closer look at the changes here later this week, but this is a promising start.

flavio committed:
http-wasmtime: handle request config inside of blocking client

Ensure blocking reqwest client uses user configuration.

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio committed:
http-wasmtime: fix building when rustls is used

Fix build time issues when rustls is being used.

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio committed:
http-wasmtime: add more logging

Log error returned by reqwest. This simplifies the debugging process.

Signed-off-by: Flavio Castelli <fcastelli@suse.com>
